river/cluster/clustream_kernel.py

```diff
@@ -0,0 +1,151 @@
+    def center(self):
+        res = {
+            i: self.linear_sum[i] / self.n_samples for i in range(len(self.linear_sum))
+        }
+        return res
+
+    def is_empty(self):
+        return self.n_samples == 0
+
+    @property
+    def radius(self):
+        if self.n_samples == 1:
+            return 0
+        return self._deviation * self.T
+
+    @property
+    def _deviation(self):
+        variance = self._variance_vector
+        sum_of_deviation = 0
+        for i in range(len(variance)):
+            d = math.sqrt(variance[i])
+            sum_of_deviation += d
+        return sum_of_deviation / len(variance)
+
+    @property
+    def _variance_vector(self):
+        res = {}
+        for i in range(len(self.linear_sum)):
+            ls = self.linear_sum[i]
+            ss = self.squared_sum[i]
+            ls_div_n = ls / self.weight
+            ls_div_n_squared = ls_div_n * ls_div_n
+            ss_div_n = ss / self.weight
+            res[i] = ss_div_n - ls_div_n_squared
+
+            if res[i] <= 0.0:
+                if res[i] > -EPSILON:
+                    res[i] = MIN_VARIANCE
+        return res
+
+    @property
+    def weight(self):
+        return self.n_samples
+
+    def insert(self, x, sample_weight, timestamp):
+        self.n_samples += sample_weight
+        self.linear_sum_timestamp += timestamp * sample_weight
+        self.squared_sum_timestamp += timestamp * sample_weight
+        for i in range(len(x)):
+            self.linear_sum[i] += x[i] * sample_weight
+            self.squared_sum[i] += x[i] * x[i] * sample_weight
+
+    @property
+    def relevance_stamp(self):
+        if self.n_samples < 2 * self.M:
+            return self._mu_time
+        return self._mu_time + self._sigma_time * self._quantile(
+            float(self.M) / (2 * self.n_samples)
+        )
+
+    @property
+    def _mu_time(self):
+        return self.linear_sum_timestamp / self.n_samples
+
+    @property
+    def _sigma_time(self):
+        return math.sqrt(
+            self.squared_sum_timestamp / self.n_samples
+            - (self.linear_sum_timestamp / self.n_samples)
+            * (self.linear_sum_timestamp / self.n_samples)
+        )
+
+    def _quantile(self, z):
+        assert 0 <= z <= 1
+        return math.sqrt(2) * self.inverse_error(2 * z - 1)
+
+    @staticmethod
+    def inverse_error(x):
+        z = math.sqrt(math.pi) * x
+        res = x / 2
+        z2 = z * z
+
+        zprod = z2 * z
+        res += (1.0 / 24) * zprod
+
+        zprod *= z2  # z5
+        res += (7.0 / 960) * zprod
+
+        zprod *= z2  # z ^ 7
+        res += (127 * zprod) / 80640
+
+        zprod *= z2  # z ^ 9
+        res += (4369 * zprod) / 11612160
+
+        zprod *= z2  # z ^ 11
+        res += (34807 * zprod) / 364953600
+
+        zprod *= z2  # z ^ 13
+        res += (20036983 * zprod) / 797058662400
+
+        return res
+
+    def add(self, cluster):
+        assert len(cluster.linear_sum) == len(self.linear_sum)
```

Comment:
`assert` must be used with caution. If really needed then as an error message so the user knows what caused the error. We must assume that not all users are experts.

Reply:
Yep I missed but we definitely don't want asserts in the code. Raise an exception instead.

Reply:
I have changed `assert` into an `if ` condition so as if a new cluster passes by while its dimension does not match with the dimension of `self`, it will do nothing instead of raising an error.